discovery/config.go

```diff
@@ -115,6 +129,53 @@ type LoginCredential struct {
 	Meta map[string]string
 }
 
+type BackOffConfig struct {
+	// InitialInterval is initial backoff retry interval for exponential backoff. Default: 500ms.
+	InitialInterval time.Duration
+	// Multiplier is the factor by which the backoff retry interval increases on each subquent
+	// retry. Default: 1.5.
+	Multiplier float64
+	// MaxInterval is the maximum backoff interval for exponential backoff. Default: 1m.
+	MaxInterval time.Duration
+	// RandomizationFactor randomizes the backoff retry interval using the formula:
+	//    RetryInterval * (random value in range [1-RandomizationFactor, 1+RandomizationFactor])
+	// Default: 0.5.
+	RandomizationFactor float64
+	// ResetInterval determines how long before resetting the backoff policy, If we are in a good
+	// state for at least this interval, then exponential backoff is reset. Default: 3m.
+	ResetInterval time.Duration
+}
+
+func (b BackOffConfig) withDefaults() BackOffConfig {
+	if b.InitialInterval == 0 {
+		b.InitialInterval = DefaultBackOffInitialInterval
+	}
+	if b.Multiplier == 0 {
+		b.Multiplier = DefaultBackOffMultiplier
+	}
+	if b.MaxInterval == 0 {
+		b.MaxInterval = DefaultBackOffMaxInterval
+	}
+	if b.RandomizationFactor == 0 {
+		b.RandomizationFactor = DefaultBackOffRandomizationFactor
+	}
+	if b.ResetInterval == 0 {
+		b.ResetInterval = DefaultBackOffResetInterval
+	}
+	return b
+}
+
+func (b BackOffConfig) getPolicy() backoff.BackOff {
+	result := backoff.NewExponentialBackOff()
+	result.InitialInterval = b.InitialInterval
+	result.MaxInterval = b.MaxInterval
+	result.Multiplier = b.Multiplier
+	result.RandomizationFactor = b.RandomizationFactor
+	// Backoff forever.
+	result.MaxElapsedTime = 0
```

Comment:
I did not allow `MaxElapsedTime` to be configurable.

This would cause the Watcher to abort when the `MaxElapsedTime` is up. If a caller wants to put a time limit on the Watcher, they can put a timeout on the context (`context.WithTimeout`).